### ISSUE TITLE
The request in signTransaction to `get_required_keys` is failing

### ISSUE THREAD

**Issue by randytorres** [EOSIO/ual-scatter#22]
I created an new create react app and added 
`"ual-reactjs-renderer": "0.1.0"`
`"ual-scatter": "^0.1.0"`

Trying to push a transaction using `activeUser.signTransaction` results in a failure. The call to `get_required_keys` fails with a 401 code even though it has the required keys. 

<img width="1162" alt="Screen Shot 2019-04-04 at 8 08 37 AM" src="https://user-images.githubusercontent.com/14025782/55567591-d76e2c80-56b2-11e9-923c-bed9d5e56808.png">

**Comment by kesar** [comment]
happening similar issue to me, but in my case available_keys are empty.

**Comment by jeffreyssmith2nd** [comment]
A good first start may be to make sure the latest version of the Scatter SDK is being used.

**Comment by randytorres** [comment]
I updated my local `ual-scatter` repo to use these package below and I had the same issue.
```
    "scatterjs-core": "2.7.16",
    "scatterjs-plugin-eosjs2": "1.5.0"
```

**Comment by eedsilva** [comment]
Same issue here with `available_keys ` empty.

**Comment by grigio** [comment]
Hi I'm having the same issue I tryied https://github.com/EOSIO/ual-reactjs-renderer and it was working, then I retried it and it doesn't work anymore

![Schermata da 2019-04-07 12-20-27](https://user-images.githubusercontent.com/8074/55682143-a4849c80-592f-11e9-8e9f-2e6bccf278cd.png)

**Comment by grigio** [comment]
News? Probably this issue is Scatter specific? https://github.com/GetScatter/scatter-js/issues/92

**Comment by chris-allnutt** [comment]
@grigio that's stating we're not authorized to hit the scatter node "nodes.get-scatter.com" . I wonder if get_required_keys is protected

**Comment by jeffreyssmith2nd** [comment]
> @grigio that's stating we're not authorized to hit the scatter node "nodes.get-scatter.com" . I wonder if get_required_keys is protected

I agree there definitely seems to be something up with the interaction between Scatter and the nodeos endpoint here. Can you post what endpoint you're using?

**Comment by kesar** [comment]
I'm getting the empty available_keys pointing from my local to https://kylin.eoscanada.com

I'm wondering if issue only happens when developing from localhost.

**Comment by randytorres** [comment]
I get the error on both localhost and when using a hosted endpoint

**Comment by randytorres** [comment]
@jeffreyssmith2nd @chris-allnutt Its also worth mentioning that the transit api seems to be having this issue as well. Theres a link to that issue in the scatter issue posted by @grigio

**Comment by grigio** [comment]
Did somebody try to upgrade the deps as suggested in #23 ?

**Comment by laurensV** [comment]
I had the same problem with the ual-plainjs-renderer with ual-scatter. Updating the eosjs dependency to 20.0.0 fixed this issue for me

**Comment by slyon** [comment]
same problem for me...

**Comment by slyon** [comment]
Getting rid of additional scatterjs version (e.g. dependecies/node_modules of your dApp's package.json) – in addition the the scatterjs dependency of `ual-scatter` seems to fix the problem, too.

**Comment by grigio** [comment]
I tried to put in `package.json`

```
  "dependencies": {
     ...
    "ual-scatter": "git://github.com/EOSIO/ual-scatter.git#develop"
  },
```

but then I get ual-scatter not found. How is possible to use #develop branch even if `ual-scatter` isn't updated yet on NPM ?

**Comment by jeffreyssmith2nd** [comment]
@grigio Did you manually add it? Doing `yarn add git://github.com/EOSIO/ual-scatter.git#develop` seems to work for me.

**Comment by grigio** [comment]
> @grigio Did you manually add it? Doing `yarn add git://github.com/EOSIO/ual-scatter.git#develop` seems to work for me.

I tried manually and with your command but I always get..

```
Module not found: Can't resolve 'ual-scatter'

❯ ls node_modules/ual-scatter/
CONTRIBUTING.md  node_modules/  README.md  tsconfig.json  yarn.lock
LICENSE          package.json   src/       tslint.json

```

**Comment by jeffreyssmith2nd** [comment]
Ah I think the issue is when you do a node_module like that, you're not going to get the compiled version. So you would have to `cd node_modules/ual-scatter && yarn && yarn build` to get the built JS to use it.

**Comment by grigio** [comment]
@jeffreyssmith2nd yeah, I noticed `dist/` was missing.. thanks!

**Comment by grigio** [comment]
Ok `ual-scatter` works but I still can't sign transactions.. Now with a different error

```
TX: eosio.token testborrow11 eosusdcom111 0.0001 EOS support

App.tsx:311 Error: Unable to sign the given transaction
    at ScatterUser._callee$ (ScatterUser.js:131)
    at tryCatch (runtime.js:63)
    at Generator.invoke [as _invoke] (runtime.js:282)
    at Generator.prototype.<computed> [as throw] (runtime.js:116)
    at rejected (ScatterUser.js:27)
```

**Comment by randytorres** [comment]
I just tried again and I'm still getting the 401 error from `get_required_keys`

**Comment by TaraTritt** [comment]
The root cause of this issue has been identified as a configuration problem within your Scatter desktop application. 

The UAL Scatter authenticator has a required field `appName` during instantiation:
```
const scatter = new Scatter([exampleNet], { appName: 'Example App' })
```

This value is used by the Scatter app to store relevant information on the application, including the `Accounts Provided`, which includes the account and permission associated with that application.
<img width="1271" alt="Screen Shot 2019-05-07 at 2 07 32 PM" src="https://user-images.githubusercontent.com/10996879/57322517-bd6ca300-70d1-11e9-931a-a5d4d91a3802.png">

During local development, often the same `appName` is reused, but the `Accounts Provided` is never updated to the latest account attempting to propose transactions. Hence an error is returned when trying to call `get_required_keys` in eosjs, which validates that the keys returned by Scatter, the stale keys, against the permission declared in the transaction, the latest account, that does not use these stale keys.

The fixes are as follows:
You can either completely delete the application in Scatter, by hovering over the application that matches the passed in `appName` and selecting the delete icon:  
<img width="612" alt="Screen Shot 2019-05-07 at 1 56 15 PM" src="https://user-images.githubusercontent.com/10996879/57322525-c2315700-70d1-11e9-87c1-b5483ce5ffa5.png">

Or if it is desired to keep that application, rename the `appName` to a unique value not already used by your other Scatter applications.

Both solutions will cause Scatter to request a reauthorization of this new application and reset the `Accounts Provided` to your latest account when attempting to propose transactions.

**Comment by TaraTritt** [comment]
To follow up on this, calling the renderer's `logout` method will also clear the application from Scatter, which should also fix this issue

**Comment by grigio** [comment]
@TaraTritt I tried to logout, delete the app in Scatter, and add a different appName, but the problem persist

**Comment by TaraTritt** [comment]
@grigio when you view the app in Scatter and select it to view its details, does the `Accounts Provided` match the account you're intending on proposing the transaction with?

**Comment by grigio** [comment]
Yes, it's supposeduser@active and supposeduser is the one which is logged in

**Comment by TaraTritt** [comment]
@grigio are you experiencing the 401 error on the `get_required_keys`request or the transaction signing error you mentioned previously? If you are having issues signing transactions, please update your package.json to use the latest published version `"ual-scatter": "^0.1.1",`. Today a patch version was published to npm

**Comment by kesar** [comment]
i am still getting it w ual-scatter ^0.1.1

**Comment by TaraTritt** [comment]
@kesar please also follow the steps in my [comment above](https://github.com/EOSIO/ual-scatter/issues/22#issuecomment-490190422) as well. Let us know if you're still having issues after updating the version and following those steps

**Comment by kesar** [comment]
No, I tried removing all the related apps in scatter and it didnt fix it. I also tried renaming the appName and it didn't work

**Comment by TaraTritt** [comment]
If you could update with your UAL configuration & Scatter configuration, omitting any private keys or any other confidential data, that would help to replicate / debug your issue would be appreciated. @randytorres confirmed that [these steps](https://github.com/EOSIO/ual-scatter/issues/22#issuecomment-490190422) resolved his problem for which this issue was originally opened. You may not have the same root cause

**Comment by kesar** [comment]
mmm its actually working, it might be some cache stuff on my end. good work guys 👍
